File: dart/test/record_test.dart
@TestOn('vm')
library;

import 'dart:io';

import 'package:servirtium/servirtium.dart';
import 'package:test/test.dart';

import 'support.dart';

void main() {
  late HttpServer upstream;
  late String upstreamBase;
  late String tapePath;

  setUp(() async {
    upstream = await HttpServer.bind('127.0.0.1', 0);
    upstreamBase = 'http://127.0.0.1:${upstream.port}';
    upstream.listen((req) async {
      const body = 'hello-from-upstream';
      req.response
        ..headers.contentType = ContentType.text
        ..contentLength = body.length; // explicit length → non-chunked
      req.response.write(body);
      await req.response.close();
    });
    tapePath = '${Directory.systemTemp.path}/vcr_dart_${DateTime.now().microsecondsSinceEpoch}.md';
  });

  tearDown(() async {
    await upstream.close(force: true);
    final f = File(tapePath);
    if (f.existsSync()) f.deleteSync();
  });

  test('records then replays the same interaction', () async {
    final rec = Vcr.record(tapePath, upstreamBase).port(0).start();
    final (_, body) = await httpGet('${rec.baseUrl}/greeting');
    expect(body, 'hello-from-upstream');
    rec.close(); // flushes the tape

    expect(File(tapePath).existsSync(), isTrue);

    final play = Vcr.playback(tapePath).port(0).start();
    try {
      final (_, replayed) = await httpGet('${play.baseUrl}/greeting');
      expect(replayed, 'hello-from-upstream');
      expect(play.lastKind, VcrOutcome.ok);
    } finally {
      play.close();
    }
  });

  test('redacts response body before it lands on the tape', () async {
    final rec = Vcr.record(tapePath, upstreamBase)
        .redact(VcrField.responseBody, 'hello-from-upstream', 'REDACTED')
        .port(0)
        .start();
    await httpGet('${rec.baseUrl}/greeting');
    rec.close();

    final tape = File(tapePath).readAsStringSync();
    expect(tape, contains('REDACTED'));
    expect(tape, isNot(contains('hello-from-upstream')));
  });
}

File: javascript/src/record.test.ts
import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import { Vcr, VcrField, VcrOutcome } from './index'
import { startFakeUpstream, startChunkedUpstream, FakeUpstream } from './test-support'

function tmpTape(): string {
  return path.join(os.tmpdir(), `vcr_rec_${Math.random().toString(36).slice(2)}.md`)
}

describe('record', () => {
  let upstream: FakeUpstream
  let tapePath: string

  afterEach(async () => {
    if (upstream) await upstream.close()
    if (tapePath && fs.existsSync(tapePath)) fs.rmSync(tapePath, { force: true })
  })

  it('records then replays the same interaction (chunked upstream de-chunked)', async () => {
    upstream = await startChunkedUpstream('hello-from-upstream')
    tapePath = tmpTape()

    // ---- record ----
    const rec = Vcr.record(tapePath, upstream.baseUrl).port(0).start()
    try {
      const res = await fetch(`${rec.baseUrl}/greeting`)
      expect(await res.text()).toBe('hello-from-upstream')
    } finally {
      rec.close() // flushes the tape
    }
    expect(fs.existsSync(tapePath)).toBe(true)
    // The recorder must store the DECODED payload, not chunk framing.
    expect(fs.readFileSync(tapePath, 'utf8')).toContain('hello-from-upstream')

    // ---- replay (offline) ----
    const play = Vcr.playback(tapePath).port(0).start()
    try {
      const replayed = await fetch(`${play.baseUrl}/greeting`)
      expect(await replayed.text()).toBe('hello-from-upstream')
      expect(play.lastKind).toBe(VcrOutcome.Ok)
    } finally {
      play.close()
    }
  })

  it('redacts a value out of the recorded tape', async () => {
    upstream = await startFakeUpstream('token=super-secret-value here')
    tapePath = tmpTape()

    const rec = Vcr.record(tapePath, upstream.baseUrl)
      .redact(VcrField.ResponseBody, 'super-secret-value', 'REDACTED')
      .port(0)
      .start()
    try {
      const res = await fetch(`${rec.baseUrl}/data`)
      // The live SUT still sees the real bytes.
      expect(await res.text()).toBe('token=super-secret-value here')
    } finally {
      rec.close()
    }

    const onDisk = fs.readFileSync(tapePath, 'utf8')
    expect(onDisk).toContain('token=REDACTED here')
    expect(onDisk).not.toContain('super-secret-value')
  })

  it('attaches a builder note to the first recorded interaction', async () => {
    upstream = await startFakeUpstream('noted-body')
    tapePath = tmpTape()

    const rec = Vcr.record(tapePath, upstream.baseUrl)
      .note('Login', 'Establishes the session the next calls reuse')
      .port(0)
      .start()
    try {
      await fetch(`${rec.baseUrl}/login`)
    } finally {
      rec.close()
    }

    const onDisk = fs.readFileSync(tapePath, 'utf8')
    expect(onDisk).toContain('Login')
    expect(onDisk).toContain('Establishes the session the next calls reuse')
  })
})
